src/vm/database/sqlite.rs

```diff
@@ -0,0 +1,305 @@
+        if !map_descriptor.key_type.admits(key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), (*key).clone())))
+        }
+
+        let params: [&ToSql; 2] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?];
+
+        let sql_result: Option<Option<String>> = 
+            self.query_row(
+                "SELECT value FROM data_table WHERE map_identifier = ? AND key = ? ORDER BY data_identifier DESC",
+                &params,
+                |row| {
+                    row.get(0)
+                })?;
+        match sql_result {
+            None => {
+                Ok(Value::Void)
+            },
+            Some(sql_result) => {
+                match sql_result {
+                    None => Ok(Value::Void),
+                    Some(value_data) => Value::deserialize(&value_data)
+                }
+            }
+        }
+    }
+
+    fn set_entry(&mut self, contract_name: &str, map_name: &str, key: Value, value: Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(&key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), key)))
+        }
+        if !map_descriptor.value_type.admits(&value) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.value_type), value)))
+        }
+
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &Some(value.serialize()?)];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
+            &params)?;
+
+        return Ok(Value::Void)
+    }
+
+    fn insert_entry(&mut self, contract_name: &str, map_name: &str, key: Value, value: Value) -> Result<Value> {
+        let map_descriptor = self.load_map(contract_name, map_name)?;
+        if !map_descriptor.key_type.admits(&key) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.key_type), key)))
+        }
+        if !map_descriptor.value_type.admits(&value) {
+            return Err(Error::new(ErrType::TypeError(format!("{:?}", map_descriptor.value_type), value)))
+        }
+
+        let exists = self.fetch_entry(contract_name, map_name, &key)? != Value::Void;
+        if exists {
+            return Ok(Value::Bool(false))
+        }
+
+        let params: [&ToSql; 3] = [&map_descriptor.map_identifier,
+                                   &key.serialize()?,
+                                   &Some(value.serialize()?)];
+
+        self.execute(
+            "INSERT INTO data_table (map_identifier, key, value) VALUES (?, ?, ?)",
+            &params)?;
+
+        return Ok(Value::Bool(true))
+    }
+
+    fn delete_entry(&mut self, contract_name: &str, map_name: &str, key: &Value) -> Result<Value> {
```

Comment:
Same as above -- should this whole operation be done in a transaction?

Reply:
Yep, same response. Since all of these functions are already being called within savepoints, I don't think there'd be any semantic difference with nesting it inside of another savepoint.